test/integration/ads_integration_test.cc

```diff
@@ -889,4 +889,236 @@ TEST_P(AdsClusterFromFileIntegrationTest, BasicTestWidsAdsEndpointLoadedFromFile
+        {buildClusterLoadAssignment("eds_cluster")}, {}, "1");
+
+    EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().Runtime, "1", {"ads_rtds_layer"}, {},
+                                        {}, false));
+    EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().Cluster, "", {}, {}, {}, false));
+    sendDiscoveryResponse<envoy::config::cluster::v3::Cluster>(
+        Config::TypeUrl::get().Cluster, {buildCluster("cluster_0")}, {buildCluster("cluster_0")},
+        {}, "1");
+
+    EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().ClusterLoadAssignment, "1",
+                                        {"eds_cluster"}, {}, {}, false));
+    EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().ClusterLoadAssignment, "1",
+                                        {"cluster_0", "eds_cluster"}, {"cluster_0"}, {}, false));
+
+    sendDiscoveryResponse<envoy::config::endpoint::v3::ClusterLoadAssignment>(
+        Config::TypeUrl::get().ClusterLoadAssignment, {buildClusterLoadAssignment("cluster_0")},
+        {buildClusterLoadAssignment("cluster_0")}, {}, "1");
+
+    EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().Cluster, "1", {}, {}, {}, false));
+    EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().Listener, "", {}, {}, {}, false));
+    sendDiscoveryResponse<envoy::config::listener::v3::Listener>(
+        Config::TypeUrl::get().Listener, {buildListener("listener_0", "route_config_0")},
+        {buildListener("listener_0", "route_config_0")}, {}, "1");
+
+    EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().ClusterLoadAssignment, "1",
+                                        {"cluster_0", "eds_cluster"}, {}, {}, false));
+    EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().RouteConfiguration, "",
+                                        {"route_config_0"}, {"route_config_0"}, {}, false));
+    sendDiscoveryResponse<envoy::config::route::v3::RouteConfiguration>(
+        Config::TypeUrl::get().RouteConfiguration,
+        {buildRouteConfig("route_config_0", "cluster_0")},
+        {buildRouteConfig("route_config_0", "cluster_0")}, {}, "1");
+
+    EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().Listener, "1", {}, {}, {}, false));
+    EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().RouteConfiguration, "1",
+                                        {"route_config_0"}, {}, {}, false));
+
+    test_server_->waitForCounterGe("listener_manager.listener_create_success", 1);
+    makeSingleRequest();
+    const ProtobufWkt::Timestamp first_active_listener_ts_1 =
+        getListenersConfigDump().dynamic_listeners(0).active_state().last_updated();
+    const ProtobufWkt::Timestamp first_active_cluster_ts_1 =
+        getClustersConfigDump().dynamic_active_clusters()[0].last_updated();
+    const ProtobufWkt::Timestamp first_route_config_ts_1 =
+        getRoutesConfigDump().dynamic_route_configs()[0].last_updated();
```

Comment:
This is also very long and hard to spot the actual code under test, can you distill it down to the essence?

Reply:
done